### PR TITLE
ARROW-5284: [Rust] Replace libc with std::alloc for memory allocation

### PR DESCRIPTION
**Pull request by sunchao**

### CODE REVIEW

File: rust/arrow/src/memory.rs [hunk 1]
```diff
@@ -18,80 +18,35 @@
 //! Defines memory-related functions, currently mostly to make this library play nicely
 //! with C.
```
Comment (andygrove) [inline rust/arrow/src/memory.rs]: is this comment still relevant?

**Review by andygrove — APPROVED** [review]
LGTM with one question.

**Comment by sunchao** [comment]
Please do not commit this yet - I still need to investigate the windows test failure.

File: rust/arrow/src/memory.rs (hunk 1 above)
Comment (sunchao) [reply rust/arrow/src/memory.rs]: This doesn't seem related anymore. Will delete later.

**Comment by nevi-me** [comment]
Hey @sunchao, I'm able to reproduce the failure, it comes from `buffer::tests::test_from_raw_parts`. specifically: `Buffer::from_raw_parts(null_mut(), 0);` causes the failure.

I ran cargo test with 1 thread to find the offender.

**Comment by sunchao** [comment]
Thanks @nevi-me . I think it fails because it was trying to dealloc a null pointer. Not sure why it doesn't fail on linux.

**Review by nevi-me — APPROVED** [review]
LGTM! Would this fix the 64-byte counting issue that I was experiencing on #4167 ?

**Comment by sunchao** [comment]
> LGTM! Would this fix the 64-byte counting issue that I was experiencing on #4167 ?

No. I'm going to work on that next.

**Pull request closed without merging**